lte/gateway/python/magma/pipelined/app/classifier.py

```diff
@@ -0,0 +1,259 @@
+
+GTP_PORT_MAC = "02:00:00:00:00:01"
+
+class Classifier(MagmaController):
+    """
+    A controller that sets up an openflow pipeline for Magma.
+
+    This controller is used for table 0 for gtp default entry and 
+    gtp tunnel.
+    """
+    APP_NAME = "classifier"
+    APP_TYPE = ControllerType.SPECIAL
+    ClassifierConfig = namedtuple(
+            'ClassifierConfig',
+            ['gtp_port', 'uplink_port_name', 'mtr_ip', 'mtr_port'],
+    )
+
+    def __init__(self, *args, **kwargs):
+        super(Classifier, self).__init__(*args, **kwargs)
+        self.config = self._get_config(kwargs['config'])
+        self.tbl_num = self._service_manager.get_table_num(self.APP_NAME)
+        self.next_table = self._service_manager.get_table_num(INGRESS)
+        self._uplink_port = OFPP_LOCAL
+        self._mtr_service_enabled = True
+        self._cookie = 1
+        self._datapath = None
+
+    def _get_config(self, config_dict):
+        port_name = None
+        mtr_ip = None
+        mtr_port = None
+        
+        if 'ovs_uplink_port_name' in config_dict:
+            port_name = config_dict['ovs_uplink_port_name']
+
+        if 'mtr_ip' in config_dict:
+            self._mtr_service_enabled = True
+            mtr_ip = config_dict['mtr_ip']
+            mtr_port = config_dict['ovs_mtr_port_number']
+        return self.ClassifierConfig(
+            gtp_port=config_dict['ovs_gtp_port_number'],
+            uplink_port_name=port_name,
+            mtr_ip=mtr_ip,
+            mtr_port=mtr_port,
+        )
+
+    def initialize_on_connect(self, datapath):
+        self._datapath = datapath
+        self._delete_all_flows()
+        self._install_default_tunnel_flows()
+
+    def _delete_all_flows(self):
+        flows.delete_all_flows_from_table(self._datapath, self.tbl_num)
+
+    def cleanup_on_disconnect(self, datapath):
+        self._delete_all_flows()
+
+    def convert_precedence_to_priority(self, precedence:int):
+        if precedence < flows.MAXIMUM_PRIORITY:
+            priority = flows.MAXIMUM_PRIORITY - precedence
+        else:
+            priority = 0
+
+        if priority < flows.DEFAULT_PRIORITY:
+            priority = flows.DEFAULT_PRIORITY
+        return priority
+
+    def _install_default_tunnel_flows(self):
+        match = MagmaMatch()
+        flows.add_flow(self._datapath,self.tbl_num, match,
+                       priority=flows.MINIMUM_PRIORITY,
+                       goto_table=self.next_table)
+
+    def _add_gtp_tunnel_flows(self, precedence:int, i_teid:int = None,
+                              o_teid:int = None, ue_ip_adr:str = None,
+                              enodeb_ip_addr:str = None, sid:int = None):
+        parser = self._datapath.ofproto_parser
+        priority = self.convert_precedence_to_priority(precedence)
+        # Add flow for gtp port
+        if i_teid:
+            match = MagmaMatch(tunnel_id=i_teid, in_port=self.config.gtp_port)
```

Comment:
I do not think we can have teid or ue-ip address NULL, so lets not make it optional.

Reply:
Done